transcode/transcode.go

```diff
@@ -165,6 +155,68 @@ func RunTranscodeProcess(transcodeRequest TranscodeSegmentRequest, streamName st
 		return outputs, segmentsCount, err
 	}
 
+	// Transmux received segments from T into a single mp4
+	if transcodeRequest.GenerateMP4 {
+		for rendition, segments := range renditionList.RenditionSegmentTable {
+			// a. create folder to hold transmux-ed files in local storage temporarily
+			err := os.MkdirAll(TRANSMUX_STORAGE_DIR, 0700)
+			if err != nil && !os.IsExist(err) {
+				log.Log(transcodeRequest.RequestID, "failed to create temp dir for transmuxing", "dir", TRANSMUX_STORAGE_DIR, "err", err)
+				return outputs, segmentsCount, err
+			}
+
+			// b. create a single .ts file for a given rendition by concatenating all segments in order
+			concatTsFileName := filepath.Join(TRANSMUX_STORAGE_DIR, transcodeRequest.RequestID+"_"+rendition+".ts")
+			totalBytes, err := video.ConcatTS(concatTsFileName, segments)
+			if err != nil {
+				log.Log(transcodeRequest.RequestID, "error concatenating .ts", "file", concatTsFileName, "err", err)
+				continue
+			}
+			defer os.Remove(concatTsFileName)
+
+			// c. Verify the total bytes written for the single .ts file for a given rendition matches the total # of bytes we received from T
+			renditionIndex := getProfileIndex(transcodeProfiles, rendition)
+			var rendBytesWritten int64 = -1
+			for _, v := range transcodedStats {
+				if v.Name == rendition {
+					rendBytesWritten = v.Bytes
+				}
+			}
+			if rendBytesWritten != totalBytes {
+				log.Log(transcodeRequest.RequestID, "bytes written does not match", "file", concatTsFileName, "bytes expected", transcodedStats[renditionIndex].Bytes, "bytes written", totalBytes)
+				break
+			}
+
+			// d. Transmux the single .ts file into an .mp4 file
+			mp4OutputFileName := concatTsFileName[:len(concatTsFileName)-len(filepath.Ext(concatTsFileName))] + ".mp4"
+			err = video.MuxTStoMP4(concatTsFileName, mp4OutputFileName)
+			if err != nil {
+				log.Log(transcodeRequest.RequestID, "error transmuxing", "err", err)
+				continue
+			}
+
+			// e. Upload mp4 output file
+			mp4OutputFile, err := os.Open(mp4OutputFileName)
+			if err != nil {
+				log.Log(transcodeRequest.RequestID, "error opening mp4", "file", mp4OutputFileName, "err", err)
+				break
+			}
+			defer os.Remove(mp4OutputFileName)
+
+			targetMP4URL, err := url.JoinPath(targetTranscodedRenditionOutputURL.String(), rendition)
+			if err != nil {
+				log.Log(transcodeRequest.RequestID, "error building url for MP4 output", "url", targetMP4URL, "err", err)
+			}
+
+			err = backoff.Retry(func() error {
+				return clients.UploadToOSURL(targetMP4URL, filepath.Base(mp4OutputFile.Name()), bufio.NewReader(mp4OutputFile), UPLOAD_TIMEOUT)
+			}, clients.UploadRetryBackoff())
+			if err != nil {
+				log.Log(transcodeRequest.RequestID, "failed to upload mp4", "file", mp4OutputFile.Name())
+			}
+		}
+	}
+
 	manifestURL = strings.ReplaceAll(manifestURL, targetTranscodedRenditionOutputURL.String(), playbackBaseURL)
 	output := video.OutputVideo{Type: "object_store", Manifest: manifestURL}
```

Comment:
Will you be adding the mp4 urls to the output struct in a follow up PR?

Reply:
Yeah this change will go in when we make catalyst the primary pipeline.